### PR TITLE
Generate base docs

### PR DESCRIPTION
**Pull request by youngkidwarrior**
This PR adds jsdoc to to the repo. Docs are saved locally in the /docs folder. 

Use the `yarn docs` command to generate docs programmatically seperated by class. Using the already available documentation.

Changes are not made to docusaurus, following PRs will handle this.

Ignoring flow typing in script file as well. I'm unfamiliar with flow, so I can handle this in another PR if it is necessary.

Here's an example of the docs running on my local machine.

Cred_Docs

### CODE REVIEW

File: scripts/build-docs.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  /* reduce templateData to an array of class names */
+  const classNames = templateData.reduce((classNames, identifier) => {
+    if (identifier.kind === "class") classNames.push(identifier.name);
+    return classNames;
+  }, []);
+
+  /* create a documentation file for each class */
+  for (const className of classNames) {
+    const template = `{{#class name="${className}"}}{{>docs}}{{/class}}`;
+    console.log(`rendering ${className}, template: ${template}`);
```
Comment (topocount) [inline scripts/build-docs.js:49]: It'd be helpful to wrap these types of log messages in a function with a good name. We make use of `chalk` to indicate the type of message getting logged to the console.

File: scripts/build-docs.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+}
+
+(async () => {
+  try {
+    getFilesFromDir("./src", [".js"]).map((fileDir) => generateDocs(fileDir));
+  } catch (e) {
+    console.error("We've thrown! Whoops!", e);
+  }
+})();
```
Comment (topocount) [inline scripts/build-docs.js:65]: IIFE's are cool, but I'd prefer to see this written as a `then`able expression.

**Review by topocount — COMMENTED** [review]
Hey @youngkidwarrior I really appreciate this contribution and the idea behind it. I think it'll be helpful to document the API since I know I learn better by reading this kind of stuff, and it'd be awesome to make a lot of the docstrings in here available to people who aren't familiar/comfortable with GitHub.

I'm not sure it makes sense to store the doc files in the master branch of this repo, however, and if they _must_ be stored in GitHub somewhere to work with docusaurus, I'd prefer that they'd be stored in the docs repo.

I imagine that in future PRs you'll be providing the integration logic with docusaurus. It'd be awesome if you could include logic that redeploys docusaurus every time JSDoc changes are pushed to master.

Finally, please work to make the build-docs script flow-friendly and not just flow-ignored. It'll make it infinitely more maintainable. I'm happy to peer up with you on this if desired.

**Comment by youngkidwarrior** [comment]
> - I'm not sure it makes sense to store the doc files in the master branch of this repo, however, and if they _must_ be stored in GitHub somewhere to work with docusaurus, I'd prefer that they'd be stored in the docs repo.

I wholeheartedly agree. The plan is to move the docs to be stored in the docs repo when I integrate it in the next PR. These are just here incase people wanted to take a look at what is generated. Should I just remove it from the PR?


> - I imagine that in future PRs you'll be providing the integration logic with docusaurus. It'd be awesome if you could include logic that redeploys docusaurus every time JSDoc changes are pushed to master.

Yup the idea is to integrate this with the CI so it updates every PR.


> - Finally, please work to make the build-docs script flow-friendly and not just flow-ignored. It'll make it infinitely more maintainable. I'm happy to peer up with you on this if desired.

Yeah I'll work this into the next PR. It will just take me a second to figure out the flow syntax compared to typescript, so I'd rather delegate that effort to a future PR. Is that okay for now?

File: scripts/build-docs.js (hunk 1 above)
Comment (youngkidwarrior) [reply scripts/build-docs.js:49]: I'm not sure I fully understand what you're asking. Do you mean extrapolate just the console log statement + dhalk out into its own function?

**Comment by topocount** [comment]
> Yeah I'll work this into the next PR. It will just take me a second to figure out the flow syntax compared to typescript, so I'd rather delegate that effort to a future PR. Is that okay for now?

Yup! It'd be best if it was it's own PR. I think it will get a lot of attention in its own right

**Review by topocount — COMMENTED** [review]
@youngkidwarrior Thanks for making these changes to remove the IIFE. The fastest way to get this particular PR merged at this point will be to delete the generated doc files in this branch and refactor the script so it works with flow.

**Review by hammadj — COMMENTED** [review]
Just ran the script and got this error:

image


Can you verify this is still working? Also, based on the GIF shown, it seems like it's not actually documenting any of the params to the functions? Is there any way we can get those working?

**Comment by youngkidwarrior** [comment]
It is working the reason it breaks for you is there is I removed the `/docs` directory as requested. 

> it's not actually documenting any of the params to the functions

Right. this was just getting jsdocs working with the current codebase. I can make those improvements in a future PR.

**Comment by hammadj** [comment]
> It is working. The reason it breaks for you is I removed the `/docs` directory as requested. Add that back and it should run fine.

@youngkidwarrior in that case, the script should automatically create the docs folder if it doesn't already exist

> Right. this was just getting jsdocs working with the current codebase. I can make those improvements in a future PR.

Cool, would also be nice to have the generated docs organized by folder. Also I noticed that it's only generating ~20ish docs files, things like `core/graph.js` or `core/address.js` dont have any generated docs. Why is that?

**Comment by youngkidwarrior** [comment]
@hammadj I have it fixed in the followup PR where I migrate to flow typing

I noticed that. I'll take a look tomorrow

**Comment by youngkidwarrior** [comment]
@hammadj Fixed the docs folder problem. The reason the other files aren't showing up is because I am separating the docs by classes. Files without classes are skipped. I'll have to play with the proper way to display all the data

**Comment by youngkidwarrior** [comment]
@hammadj @topocount It seems like I might have merged some extra commits in this. Would one of you want to pair with me to help me make sure these are solved correctly?

File: src/plugins/discourse/createGraph.js [hunk 3]
```diff
@@ -22,6 +22,8 @@ import * as NE from "./nodesAndEdges";
 import {type DiscourseConfig} from "./config";
 import {likeWeight} from "./weights";
 
+import {likeWeight} from "./weights";
+
```
Comment (hammadj) [inline src/plugins/discourse/createGraph.js:26]: This is imported twice here

**Review by hammadj — COMMENTED** [review]
Still have the DS_Store file here, should delete it and add it to git-ignore

**Review by hammadj — APPROVED** [review]
Looks good! Make sure to squash merge when you do so that the extra commits dont show up in master :)

**Comment by youngkidwarrior** [comment]
@hammadj @decentralion  looks like I'm not able to merge onto a protected brach

**Pull request merged**